[Oxidize BasisTranslator]: Move `basis_search` and `BasisSearchVisitor` to rust.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Builds on #12585
The following commits add a rust-side version of `basis_search` that is also exposed to `BasisTranslator` in Python. Due to structural restrictions in `rustworkx-core`'s version of `dijkstra_search`, these commits also consolidate the `basis_search` method and the `BasisSearchVisitor` into one method.

### Details and comments
The `BasisTranslator` aims to map a circuit's gates into a specified target basis by performing a dijkstra search in a provided `EquivalenceLibrary`. Because of #12585 moving said structure mainly to rust, these commits aim to add the `basis_search` method of this transpiler pass to operate in rust space as well, which could save the overhead of converting the rust `EquivalenceLibrary`'s `petgraph::StableDiGraph` to a `PyDiGraph`.

It is worth to note that the following changes have been made:
- `BasisSearchVisitor` and `_basis_search` have been consolidated into one method `basis_search`, due to the visitor's data living in a different space, which would imply moving some of the `rustworkx-core`'s `dijkstra_search` retrieved graph data moving outside of its expected lifetimes.
- `basis_seach` is an entirely rust-based method. However, it is exposed temporarily to Python via `py_basis_search` to avoid cloning of `source_basis` and `target_basis` sets. This exposure however will be removed once #12246 is fully realized.

### Known issues:
~~When running the test `test.python.transpiler.test_basis_translator.TestBasisExamples.test_cx_bell_to_ecr` there's a posibility that the resulting circuit will randomly be:~~

<pre style="word-wrap: normal;white-space: pre;background: #fff0;line-height: 1.1;font-family: &quot;Courier New&quot;,Courier,monospace">        ┌────────────┐   ┌─────────────┐┌──────┐┌──────────┐
q_0: ───┤ U(π/2,0,π) ├───┤ U(0,0,-π/2) ├┤0     ├┤ U(π,0,π) ├
     ┌──┴────────────┴──┐└─────────────┘│  Ecr │└──────────┘
q_1: ┤ U(-π/2,-π/2,π/2) ├───────────────┤1     ├────────────
     └──────────────────┘               └──────┘            </pre>

~~or:~~
<pre style="word-wrap: normal;white-space: pre;background: #fff0;line-height: 1.1;font-family: &quot;Courier New&quot;,Courier,monospace">        ┌────────────┐  ┌─────────────┐┌──────────┐┌──────┐
q_0: ───┤ U(π/2,0,π) ├──┤ U(0,0,-π/2) ├┤ U(π,0,0) ├┤0     ├
     ┌──┴────────────┴─┐└─────────────┘└──────────┘│  Ecr │
q_1: ┤ U(π/2,-π/2,π/2) ├───────────────────────────┤1     ├
     └─────────────────┘                           └──────┘</pre>

~~This is suspectedly due to `rustworkx-core` not being able to decide between these two equivalent alternatives.~~ Fixed by [e7b3cb0](https://github.com/Qiskit/qiskit/pull/12585/commits/e7b3cb09c9841ff2f91a2d0af625ef249a766da1)

### Blockers:
- [x] #12585

This branch will be rebased once each blocker merges. Feel free to leave any comments or review. Thank you :rocket:

raynelfss committed:
Add: Basis search function
- Add rust counterpart for `basis_search`.
- Consolidated the `BasisSearchVisitor` into the function due to differences in rust behavior.

raynelfss committed:
Fix: Wrong return value for `basis_search`

raynelfss committed:
Fix: Remove `IndexMap` and duplicate declarations.

raynelfss committed:
Fix: Adapt to #12730

raynelfss committed:
Remove: unused imports

raynelfss commented:
Although the performance gains are very niche, here are some benchmarks:
```zsh
All benchmarks:

| Change   | Before [ffead590] <move-target~1^2>   | After [b6d36cf1] <move-basis-search-and-visit>   | Ratio   | Benchmark (Parameter)                                                                                     |
|----------|---------------------------------------|--------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------|
|          | 8.20±0s                               | 6.84±0.1s                                        | ~0.83   | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([0, 1]) |
|          | 8.19±0s                               | 6.70±0.02s                                       | ~0.82   | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([0, 1])               |
|          | 111±0.3ms                             | 113±0.4ms                                        | 1.02    | quantum_info.RandomCliffordBench.time_random_clifford('4,1500')                                           |
|          | 126±0.8ms                             | 126±0.3ms                                        | 1.00    | quantum_info.RandomCliffordBench.time_random_clifford('3,2000')                                           |
|          | 104±0.2ms                             | 104±0.2ms                                        | 1.00    | quantum_info.RandomCliffordBench.time_random_clifford('5,1000')                                           |
|          | 72.8±0.4ms                            | 72.9±0.03ms                                      | 1.00    | quantum_info.RandomCnotDihedralBench.time_random_cnotdihedral('1,2000')                                   |
|          | 46.1±0.06ms                           | 46.2±0.2ms                                       | 1.00    | quantum_info.RandomCnotDihedralBench.time_random_cnotdihedral('3,1200')                                   |
|          | 39.5±0.07ms                           | 39.3±0.04ms                                      | 1.00    | quantum_info.RandomCnotDihedralBench.time_random_cnotdihedral('4,1000')                                   |
|          | 32.3±0.04ms                           | 32.3±0.2ms                                       | 1.00    | quantum_info.RandomCnotDihedralBench.time_random_cnotdihedral('5,800')                                    |
|          | 29.5±0.07ms                           | 29.3±0.4ms                                       | 1.00    | quantum_info.RandomCnotDihedralBench.time_random_cnotdihedral('6,700')                                    |
|          | 17.9±0.4ms                            | 18.0±0.2ms                                       | 1.00    | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(10)                                  |
|          | 41                                    | 41                                               | 1.00    | random_circuit_hex.BenchRandomCircuitHex.track_depth_ibmq_backend_transpile(10)                           |
|          | 49                                    | 49                                               | 1.00    | random_circuit_hex.BenchRandomCircuitHex.track_depth_ibmq_backend_transpile(12)                           |
|          | 169                                   | 169                                              | 1.00    | random_circuit_hex.BenchRandomCircuitHex.track_depth_ibmq_backend_transpile(14)                           |
|          | 17                                    | 17                                               | 1.00    | random_circuit_hex.BenchRandomCircuitHex.track_depth_ibmq_backend_transpile(4)                            |
|          | 25                                    | 25                                               | 1.00    | random_circuit_hex.BenchRandomCircuitHex.track_depth_ibmq_backend_transpile(6)                            |
|          | 33                                    | 33                                               | 1.00    | random_circuit_hex.BenchRandomCircuitHex.track_depth_ibmq_backend_transpile(8)                            |
|          | 144±0.3ms                             | 143±0.7ms                                        | 0.99    | quantum_info.RandomCliffordBench.time_random_clifford('1,3000')                                           |
|          | 133±1ms                               | 132±0.5ms                                        | 0.99    | quantum_info.RandomCliffordBench.time_random_clifford('2,2500')                                           |
|          | 80.5±0.5ms                            | 79.8±0.7ms                                       | 0.99    | quantum_info.RandomCliffordBench.time_random_clifford('6,700')                                            |
|          | 56.6±0.2ms                            | 56.2±0.07ms                                      | 0.99    | quantum_info.RandomCnotDihedralBench.time_random_cnotdihedral('2,1500')                                   |
|          | 24.4±0.2ms                            | 24.2±0.3ms                                       | 0.99    | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(12)                                  |
|          | 37.8±0.4ms                            | 37.3±0.2ms                                       | 0.99    | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(14)                                  |
|          | 8.72±0.1ms                            | 8.55±0.2ms                                       | 0.98    | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(6)                                   |
|          | 13.2±0.2ms                            | 12.9±0.07ms                                      | 0.98    | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(8)                                   |
|          | 5.93±0.1ms                            | 5.63±0.1ms                                       | 0.95    | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(4)                                   |
| -        | 4.20±0s                               | 3.25±0s                                          | 0.77    | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([0])                  |
| -        | 4.21±0s                               | 3.24±0s                                          | 0.77    | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([0])    |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
These changes seem to perform best without multithreading as serializing objects from rust to python is a bit complex at times.

raynelfss committed:
Docs: Edit docstring for rust native `basis_search`

raynelfss committed:
Fix: Use owned Strings.
- Due to the nature of `hashbrown` we must use owned Strings instead of `&str`.

raynelfss committed:
Add: mutable graph view that the `BasisTranslator` can access in Rust.
- Remove import of `random` in `basis_translator`.

raynelfss committed:
Merge branch 'main' into move-basis-search-and-visit

raynelfss committed:
Merge branch 'main' into move-basis-search-and-visit

mtreinish commented on crates/accelerate/src/equivalence.rs:
```diff
@@ -697,6 +697,11 @@ impl EquivalenceLibrary {
     }
+
+    /// Expose a mutable view of the inner graph.
+    pub(crate) fn mut_graph(&mut self) -> &mut GraphType {
```
I'd probably call this `graph_mut`, it seems to be more of the rust convention to do `something_mut` instead of `mut_something`. Thinking like `iter_mut()` vs `iter()`.

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,228 @@
+        _ => None,
+    };
+
+    // TODO: Values will have to be cloned in order for the dummy node to be removed.
```
Is this still relevant?

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,228 @@
+    py: Python,
+    equiv_lib: &mut EquivalenceLibrary,
+    source_basis: HashSet<GateIdentifier>,
+    target_basis: HashSet<String>,
```
Can't you just do `HashSet<&str>` here? I thought there was `FromPyObject` for &str.

raynelfss replied:
Apparently it isn't doable:
```
the trait bound `hashbrown::HashSet<(&str, u32)>: PyFunctionArgument<'_, '_>` is not satisfied
the trait `pyo3::FromPyObject<'py>` is implemented for `hashbrown::HashSet<K, S>`
required for `hashbrown::HashSet<(&str, u32)>` to implement `FromPyObjectBound<'_, '_>`
required for `hashbrown::HashSet<(&str, u32)>` to implement `PyFunctionArgument<'_, '_>`
```

Although this was initially designed to use `HashSet<&str>`, I will change it to use String to avoid multiple conversions with the rust `BasisTranslator` since the target basis would use owned strings to store them within the struct itself. Having this use `HashSet<&str>` would require us to iterate through the original, call `.as_string()` and collecting them into a `HashSet`. Reveiving a `&HashSet<String>` would make it just a bit easier here.

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,228 @@
+    // Initialize visitor attributes:
+    initialize_num_gates_remain_for_rule(equiv_lib.graph(), &mut num_gates_remaining_for_rule);
+
+    // TODO: Logs
```
You can just remove this we don't need to bother with pylogging here. The basis search debug log messages weren't really ever useful, even for debugging the basis translator.

```suggestion
```

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,228 @@
+    // Connect each edge in the target_basis to the dummy node.
+    target_basis_indices.iter().for_each(|node| {
+        equiv_lib.mut_graph().add_edge(dummy, *node, None);
+    });
```
As opposed to? :)

```suggestion
    for node in target_basis_indices {
        equiv_lib.mut_graph().add_edge(dummy, *node, None);
    }
```

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,228 @@
+                        borrowed_cost_map.insert(gate.clone(), score);
+                    }
+                    if let Some(rule) = predecessors.borrow().get(&gate) {
+                        // TODO: Logger
```
```suggestion
```

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,228 @@
+    source: &mut HashMap<usize, usize>,
+) {
+    let mut save_index = usize::MAX;
+    for edge_data in graph.edge_weights().flatten() {
```
Maybe leave a comment here that `flatten()` is for iterating over defined edges and skipping any set with `None`. It took me a second to figure out what this was doing, it's a clever workaround though (`Option<T>` implementing `Iterator` is always confusing to me at first).

mtreinish reviewed (commented):
Overall this LGTM, thanks for doing this. I left a few minor comments inline but overall this is a pretty straightforward porting of the Python search algorithm into Rust. I'm just glad we had that rust search interface in rustworkx-core as it makes this quite simply to move into rust.

raynelfss committed:
Fix: Review comments
- Rename `EquivalenceLibrary`'s `mut_graph` method to `graph_mut` to keep consistent with rust naming conventions.
- Use `&HashSet<String>` instead of `HashSet<&str>` to avoid extra conversion.
- Use `u32::MAX` as num_qubits for dummy node.
- Use for loop instead of foreachj to add edges to dummy node.
- Add comment explaining usage of flatten in `initialize_num_gates_remain_for_rule`.
- Remove stale comments.

mtreinish commented on crates/accelerate/src/basis/basis_translator/basis_search.rs:
```diff
@@ -0,0 +1,217 @@
+///     Optional[List[Tuple[gate, equiv_params, equiv_circuit]]]: List of (gate,
+///         equiv_params, equiv_circuit) tuples tuples which, if applied in order
+///         will map from source_basis to target_basis. Returns None if no path
+///         was found.
```
Nit: I think it's better to put the annotation on the actual function definition instead of the docstring.
```suggestion
/// Search for a set of transformations from source_basis to target_basis.
/// Args:
///     equiv_lib (EquivalenceLibrary): Source of valid translations
///     source_basis (Set[Tuple[gate_name: str, gate_num_qubits: int]]): Starting basis.
///     target_basis (Set[gate_name: str]): Target basis.
///
/// Returns:
///     Optional[List[Tuple[gate, equiv_params, equiv_circuit]]]: List of (gate,
///         equiv_params, equiv_circuit) tuples tuples which, if applied in order
///         will map from source_basis to target_basis. Returns None if no path
///         was found.
#[pyfunction]
#[pyo3(name = "basis_search")]
```

mtreinish reviewed (dismissed):
I missed this in my earlier review but the directory structure here is a bit weird relative to the rest of the modules in the crate. The double nesting of `basis/basis_translator` seems like 1 too many. However, this directory structure is already in use from the `compose_transforms()` PR. We can clean this up in a follow up PR. Realistically we're going to want to reorganize things in the near future to have a dedicated transpiler pass module or crate.

That all being said I left one nit inline. I'll apply the suggestion and reapprove then enqueue for merge.

mtreinish committed:
Update crates/accelerate/src/basis/basis_translator/basis_search.rs
